### PR TITLE
Added DevKit Syscalls 537 and 589

### PR DESCRIPTION
**Pull request by VocalFan**
Implemented the syscalls: sys_dl_notify_event and sys_dynlib_dlopen

Added them simply as they are Syscalls that immediately exit with ENOSYS on DevKit firmware.

### CODE REVIEW

File: src/kernel/src/syscalls/mod.rs [hunk 1]
```diff
@@ -75,6 +75,8 @@ impl Syscalls {
                 i.args[3].into(),
                 i.args[4].into(),
             ),
+            537 => self.devkit_syscall(), //sys_dl_notify_event
```
Comment (ultimaweapon) [inline src/kernel/src/syscalls/mod.rs:78]:
```c
undefined8 FUN_ffffffffa0411a30(long param_1)

{
    long lVar1;
    long lVar2;
    undefined local_a8 [16];
    undefined4 local_98;
    undefined4 local_90;
    long local_38;
    
    local_38 = __stack_chk_guard[0];
    lVar2 = *(long *)(param_1 + 8);
    lVar1 = lVar2 + 0xf8;
    mtx_lock(lVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0xd2f);
    bzero(local_a8,0x70);
    local_98 = 0xc;
    local_90 = 0x10006;
    FUN_ffffffffa040fec0(lVar2,0,0xc,local_a8);
    mtx_unlock(lVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0xd31);
    if (__stack_chk_guard[0] == local_38) {
        return 0x4e;
    }
// WARNING: Subroutine does not return
    __stack_chk_fail();
}
```

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/syscalls/mod.rs (hunk 1 above)
Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: Uh, and?

Sysent 537 runs this:

```
           int sy_c  nosys                   sy_call
```
And Sysent 589 runs this function:

```
undefined8 FUN_ffffffffa0624970(void)

{
  return 0x4e;
}
```

...That just returns ENOENT.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: And as I said uh... Both syscalls directly or indirectly ENOSYS...

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: Like... Looking at both Syscalls, can't find that function.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: Ah, it's ENOSYS itself.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: Wait, but you already use ENOSYS once elsewhere!

Comment (ultimaweapon) [reply src/kernel/src/syscalls/mod.rs:78]: The problem is does a call to `FUN_ffffffffa040fec0` on 537 cause any side effect or not (e.g. change process state).

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: So, you can[ implement a usage of ENOSYS](https://github.com/obhq/obliteration/blob/99dac6e5ce66eed3a65de7a4b258619dca9ade47/src/kernel/src/syscalls/mod.rs#L433C1-L435), but I'm the one that has to check when I try to use it?

Comment (ultimaweapon) [reply src/kernel/src/syscalls/mod.rs:78]: Check on that syscall and you will see the reason. I did not said you can't use `ENOSYS`. What I said is you need to understand what a call to `FUN_ffffffffa040fec0` is doing.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]:
```cpp
int FUN_ffffffffa040fec0
              (proc *param_1,thread *param_2,uint param_3,proc *param_4,undefined8 param_5,
              undefined8 param_6)

{
  mtx *pmVar1;
  int iVar2;
  uint *puVar3;
  byte *pbVar4;
  uint uVar5;
  sigacts *psVar6;
  char cVar7;
  uint uVar8;
  uint uVar9;
  int iVar10;
  sigqueue *psVar11;
  undefined8 uVar12;
  long lVar13;
  thread *ptVar14;
  uint uVar15;
  uint uVar16;
  thread *ptVar17;
  thread **in_GS_OFFSET;
  undefined8 local_88;
  undefined8 local_80;
  undefined8 local_78;
  undefined8 local_70;
  undefined8 local_68;
  undefined8 local_60;
  undefined8 local_58;
  undefined8 *local_50;
  undefined8 local_48;
  undefined4 local_40;
  long local_38;
  
  uVar16 = param_3 - 1;
  local_38 = __stack_chk_guard[0];
  if (0x7f < uVar16) {
    FUN_ffffffffa0b50490();
    FUN_ffffffffa0b50b80(0xffffffffa0410a1c,"tdsendsignal",(long)(int)param_3);
    panic(0x2b,"%s(): invalid signal %d","tdsendsignal",param_3);
    do {
      invalidInstructionException();
    } while( true );
  }
  if (param_1->p_state == 2) {
    iVar10 = 0;
    if (param_4 == (proc *)0x0) goto LAB_ffffffffa04103b0;
    uVar16 = *(uint *)&param_4->field9_0x60;
  }
  else {
    uVar8 = param_3;
    if ((*(byte *)((long)&param_1->p_flag + 1) & 8) == 0) {
      uVar8 = 9;
      if ((param_1->field_0x432 & 2) == 0) {
        uVar8 = param_3;
      }
      uVar16 = uVar8 - 1;
    }
    psVar6 = param_1->p_sigacts;
    KNOTE_LOCKED(&param_1->p_klist,(long)(int)(uVar8 | 0x8000000),1,(ulong)(uVar8 | 0x8000000),
                 param_5,param_6);
    uVar9 = 0;
    if (uVar16 < 0x1f) {
      uVar9 = *(uint *)(&DAT_ffffffffa0b5bc40 + (ulong)uVar16 * 4);
    }
    iVar10 = (int)uVar16 >> 5;
    if (param_2 == (thread *)0x0) {
      if (uVar8 == 9) {
        for (ptVar14 = param_1->p_threads_first; ptVar14 != (thread *)0x0;
            ptVar14 = (thread *)ptVar14->field2_0x10) {
          *(byte *)&ptVar14->td_dbgflags = *(byte *)&ptVar14->td_dbgflags & 0xfe;
        }
      }
      param_2 = *in_GS_OFFSET;
      if ((param_2->td_proc != param_1) ||
         (((param_2->td_sigmask).bits[iVar10] >> (uVar16 & 0x1f) & 1) != 0)) {
        ptVar14 = param_1->p_threads_first;
        if (ptVar14 == (thread *)0x0) {
          param_2 = (thread *)0x0;
        }
        else {
          ptVar17 = ptVar14;
          do {
            if (((((ptVar17->td_sigmask).bits[iVar10] & 1 << ((byte)uVar16 & 0x1f)) == 0) &&
                ((*(byte *)&ptVar17->td_dbgflags & 1) == 0)) &&
               (param_2 = ptVar17, (*(byte *)&ptVar17->td_inhibitors & 0x20) == 0)) break;
            ptVar17 = (thread *)ptVar17->field2_0x10;
            param_2 = ptVar14;
          } while (ptVar17 != (thread *)0x0);
        }
      }
      psVar11 = &param_1->p_sigqueue;
    }
    else {
      psVar11 = &param_2->td_sigqueue;
    }
    if (DAT_ffffffffa190b7c8 != 0) {
      (*DAT_ffffffffa1913ea8)(DAT_ffffffffa190b7c8,param_2,param_1,(long)(int)uVar8,0,0);
    }
    pmVar1 = &psVar6->field409_0xca8;
    mtx_lock(pmVar1);
    uVar15 = 1 << ((byte)uVar16 & 0x1f);
    lVar13 = (long)iVar10;
    if ((*(uint *)((long)&psVar6->field394_0xc50 + lVar13 * 4) >> (uVar16 & 0x1f) & 1) == 0) {
      cVar7 = '\x03';
      if (((param_2->td_sigmask).bits[lVar13] & uVar15) == 0) {
        cVar7 = ((*(uint *)((long)&psVar6->field396_0xc60 + lVar13 * 4) & uVar15) != 0) * '\x02';
      }
      uVar16 = *(uint *)((long)&psVar6->field386_0xc10 + lVar13 * 4);
      mtx_unlock(pmVar1);
      if ((uVar9 & 0x20) == 0) {
        if ((uVar9 & 4) == 0) goto LAB_ffffffffa0410388;
        if ((((uVar9 & 8) == 0) || (cVar7 != '\0')) ||
           (*(int *)((long)&param_1->p_pgrp[1].field1_0x8 + 4) != 0)) {
          local_50 = &local_58;
          local_80 = 0;
          local_78 = 0;
          local_70 = 0;
          local_68 = 0;
          local_60 = 0;
          local_58 = 0;
          local_88 = 0x40000;
          local_48 = 0;
          local_40 = 1;
          FUN_ffffffffa0412010(&param_1->p_sigqueue,&local_78,&local_88);
          ptVar14 = param_1->p_threads_first;
          if (ptVar14 != (thread *)0x0) {
            do {
              FUN_ffffffffa0412010(&ptVar14->td_sigqueue,&local_78,&local_88);
              ptVar14 = (thread *)ptVar14->field2_0x10;
            } while (ptVar14 != (thread *)0x0);
          }
          FUN_ffffffffa040bfd0(&local_78);
          if ((param_1->p_flag & 0x10000U) != 0) {
            param_1->p_flag = param_1->p_flag & 0xfffeffff;
            mtx_lock(&param_1->p_pptr->p_mtx);
            FUN_ffffffffa040beb0(param_1->p_ksi);
            mtx_unlock(&param_1->p_pptr->p_mtx);
          }
          goto LAB_ffffffffa0410388;
        }
        if (param_4 != (proc *)0x0) {
          uVar16 = *(uint *)&param_4->field9_0x60;
          goto joined_r0xffffffffa0410139;
        }
      }
      else {
        local_50 = &local_58;
        local_80 = 0;
        local_78 = 0;
        local_70 = 0;
        local_68 = 0;
        local_60 = 0;
        local_58 = 0;
        local_88 = 0x330000;
        local_48 = 0;
        local_40 = 1;
        FUN_ffffffffa0412010(&param_1->p_sigqueue,&local_78,&local_88);
        ptVar14 = param_1->p_threads_first;
        if (ptVar14 != (thread *)0x0) {
          do {
            FUN_ffffffffa0412010(&ptVar14->td_sigqueue,&local_78,&local_88);
            ptVar14 = (thread *)ptVar14->field2_0x10;
          } while (ptVar14 != (thread *)0x0);
        }
        FUN_ffffffffa040bfd0(&local_78);
LAB_ffffffffa0410388:
        iVar10 = FUN_ffffffffa0410d50(psVar11,uVar8,param_4);
        if (iVar10 != 0) goto LAB_ffffffffa04103b0;
        FUN_ffffffffa040ca20(param_2);
        if (cVar7 == '\x03') {
          iVar10 = 0;
          if (((uVar9 & 0x20) == 0) ||
             (iVar10 = 0, (*(byte *)((long)&param_1->p_flag + 2) & 2) == 0))
          goto LAB_ffffffffa04103b0;
        }
        if (uVar8 == 9) {
          *(undefined4 *)&param_1->field_0x2e0 = 0;
          *(undefined *)&param_1->field409_0x2e8 = 0;
          wakeup(&param_1->field409_0x2e8);
        }
        if ((param_1->field_0x430 & 2) == 0) {
          pbVar4 = (byte *)((long)&param_1->p_flag + 2);
          *pbVar4 = *pbVar4 & 0xfb;
        }
        if (uVar8 == 9) {
          for (ptVar14 = param_1->p_threads_first; ptVar14 != (thread *)0x0;
              ptVar14 = (thread *)ptVar14->field2_0x10) {
            thread_lock(ptVar14);
            FUN_ffffffffa03ed310(ptVar14);
            FUN_ffffffffa06d74a0
                      (ptVar14->field0_0x0,0,
                       "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x8c8);
          }
        }
        uVar5 = param_1->p_flag;
        iVar2 = (uint)((uVar16 & uVar15) != 0) * 5 + -1;
        if ((uVar5 & 0xe0000) == 0) {
          if (param_1->p_state != 1) {
            local_78 = 0;
            local_70 = 0;
            puVar3 = (uint *)((long)&local_78 + lVar13 * 4);
            *puVar3 = *puVar3 | uVar15;
            FUN_ffffffffa040c140(psVar11,&local_78);
            iVar10 = 0;
            goto LAB_ffffffffa04103b0;
          }
          if ((cVar7 == '\x02') || ((uVar5 & 0x800) != 0)) {
            FUN_ffffffffa0410f30(param_2,uVar8,cVar7,iVar2);
            iVar10 = 0;
            goto LAB_ffffffffa04103b0;
          }
          if ((uVar9 & 4) != 0) {
            iVar10 = 0;
            if ((uVar5 & 0x10) != 0) goto LAB_ffffffffa04103b0;
            param_1->p_flag = uVar5 | 0x20000;
            param_1->p_xstat = (ushort)uVar8;
            mtx_lock_spin(&param_1->p_slock,0,
                          "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x94e);
            sig_suspend_threads(param_2,param_1,1);
            if (param_1->p_numthreads == param_1->p_suspcount) {
              FUN_ffffffffa0411450(param_1);
              FUN_ffffffffa06d74a0
                        (&param_1->p_slock,0,
                         "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x959);
              FUN_ffffffffa040c2b0(param_1,param_1->p_xstat);
              goto LAB_ffffffffa0410981;
            }
            uVar12 = 0x95c;
            goto LAB_ffffffffa041097c;
          }
LAB_ffffffffa041090f:
          FUN_ffffffffa0410f30(param_2,uVar8,cVar7,iVar2);
          mtx_lock_spin(&param_1->p_slock,0,
                        "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x96b);
          if (uVar8 == 9) {
            FUN_ffffffffa0867720(param_1);
            FUN_ffffffffa0867640(param_1,param_2);
          }
          else {
            FUN_ffffffffa0867540(param_1);
          }
          uVar12 = 0x971;
        }
        else {
          if (uVar8 == 9) {
            iVar10 = 0;
            if ((uVar5 & 0x800) != 0) goto LAB_ffffffffa04103b0;
            param_1->p_flag = uVar5 & 0xfffdffff;
            goto LAB_ffffffffa041090f;
          }
          if ((uVar9 & 0x20) != 0) {
            if (((uVar5 & 0x800) != 0) && (iVar10 = 0, (*(ulong *)&param_1->field_0x430 & 2) != 0))
            goto LAB_ffffffffa04103b0;
            pmVar1 = &param_1->p_slock;
            param_1->p_flag = uVar5 & 0xfffdffff;
            mtx_lock_spin(pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x8fa
                         );
            if (param_1->p_numthreads == param_1->p_suspcount) {
              FUN_ffffffffa06d74a0
                        (pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x8fc)
              ;
              pbVar4 = (byte *)((long)&param_1->p_flag + 2);
              *pbVar4 = *pbVar4 | 1;
              param_1->p_xstat = 0x13;
              mtx_lock(&param_1->p_pptr->p_mtx);
              FUN_ffffffffa0411830(param_1,6,0x13);
              mtx_unlock(&param_1->p_pptr->p_mtx);
              mtx_lock_spin(pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",
                            0x902);
            }
            if (cVar7 == '\x02') {
              FUN_ffffffffa06d74a0
                        (pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x90f)
              ;
              goto LAB_ffffffffa041090f;
            }
            if (cVar7 != '\0') {
              FUN_ffffffffa0867540(param_1);
              FUN_ffffffffa06d74a0
                        (pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x916)
              ;
              iVar10 = 0;
              goto LAB_ffffffffa04103b0;
            }
            FUN_ffffffffa0867540(param_1);
            FUN_ffffffffa06d74a0
                      (pmVar1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x906);
            local_78 = 0;
            local_70 = 0;
            puVar3 = (uint *)((long)&local_78 + lVar13 * 4);
            *puVar3 = *puVar3 | uVar15;
LAB_ffffffffa04108b2:
            FUN_ffffffffa040c140(psVar11,&local_78);
            iVar10 = 0;
            goto LAB_ffffffffa04103b0;
          }
          if ((uVar9 & 4) != 0) {
            iVar10 = 0;
            if ((uVar5 & 0x800) != 0) goto LAB_ffffffffa04103b0;
            param_1->p_flag = uVar5 | 0x20000;
            local_78 = 0;
            local_70 = 0;
            puVar3 = (uint *)((long)&local_78 + lVar13 * 4);
            *puVar3 = *puVar3 | uVar15;
            goto LAB_ffffffffa04108b2;
          }
          mtx_lock_spin(&param_1->p_slock,0,
                        "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x934);
          thread_lock(param_2);
          if (param_2->td_wchan == (void *)0x0) {
            lVar13 = param_2->field0_0x0;
          }
          else {
            if ((*(byte *)&param_2->td_flags & 8) != 0) {
              iVar10 = FUN_ffffffffa0543190(param_2,iVar2);
              FUN_ffffffffa06d74a0
                        (param_2->field0_0x0,0,
                         "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x938);
              FUN_ffffffffa06d74a0
                        (&param_1->p_slock,0,
                         "W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x939);
              if (iVar10 != 0) {
                FUN_ffffffffa05bcd70();
              }
              goto LAB_ffffffffa0410981;
            }
            lVar13 = param_2->field0_0x0;
          }
          FUN_ffffffffa06d74a0
                    (lVar13,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x938);
          uVar12 = 0x939;
        }
LAB_ffffffffa041097c:
        FUN_ffffffffa06d74a0
                  (&param_1->p_slock,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",
                   uVar12);
      }
LAB_ffffffffa0410981:
      iVar10 = 0;
      goto LAB_ffffffffa04103b0;
    }
    if (DAT_ffffffffa190b978 != 0) {
      (*DAT_ffffffffa1913ea8)(DAT_ffffffffa190b978,param_2,param_1,(long)(int)uVar8,0,0);
    }
    mtx_unlock(pmVar1);
    iVar10 = 0;
    if (param_4 == (proc *)0x0) goto LAB_ffffffffa04103b0;
    uVar16 = *(uint *)&param_4->field9_0x60;
  }
joined_r0xffffffffa0410139:
  iVar10 = 0;
  if ((uVar16 & 6) == 4) {
    uma_zfree(DAT_ffffffffa1f10cf8,param_4,0);
    iVar10 = 0;
  }
LAB_ffffffffa04103b0:
  if (__stack_chk_guard[0] == local_38) {
    return iVar10;
  }
                    /* WARNING: Subroutine does not return */
  __stack_chk_fail();
  ```

Comment (ultimaweapon) [reply src/kernel/src/syscalls/mod.rs:78]: You have 3 choices here:

1. Understand what this function does.
2. Test if return `ENOSYS` is working without consequence.
3. Remove this implementation for now if 1 or 2 is not feasible yet.

For 2 you need to do an actual test to see if a game (or firmware) can continue to run without crashing.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: I'll look at the IDA Pro disassembly as it also has ORBIS and Unity SDK disassembled and also a bit more... Implemented value names haha.

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: There was no custom function for ENOSYS in the disassembled code (Even tried just searching `78LL`, only saw times ENOSYS was returned or used as a variable)

Comment (ultimaweapon) [reply src/kernel/src/syscalls/mod.rs:78]: What do you mean?

Comment (VocalFan) [reply src/kernel/src/syscalls/mod.rs:78]: Basically I mean the ENOSYS implementation is most probable to be the native FreeBSD version of it.

Comment (ultimaweapon) [reply src/kernel/src/syscalls/mod.rs:78]: Do you mean `FUN_ffffffffa040fec0`? If yes what it does here?

**Comment by ultimaweapon** [comment]
I'll close this for now due to no activities. You can reopen this or create a new one when you are come back to work on this.

**Pull request closed without merging**